### DISCUSSION TITLE
ISR mode, serving static files, and server's modes cleanup

### DISCUSSION THREAD

**Discussion by Nayte91** [vercel/next.js#58004] [Help]
### Summary

Hello,

I'm doing a small web project: my back is a PHP RESTful data API, and my front is a NextJS14 site with basically 2 pages (an index to find articles, and a display article one. Dynamic of course, as there is dozens of articles and new ones can appears anytime). My 2 sides are on my own server, dockerized with a reverse proxy (Traefik), and ideally a Nginx webserver then PHP or node behind.

As I really liked SSG mode for speed and SEO considerations, but as I need to dynamically manage my number of article pages, I would try to deploy with ISR mode.. But there is few questions that I can't find any answer to.

1. If I put `output: export` in my next.config.js file, then I can't use `npm start` anymore:
```bash
    > next start
    
       ▲ Next.js 14.0.1
       - Local:        http://localhost:3000
    
    Error: "next start" does not work with "output: export" configuration. Use "npx serve@latest out" instead. 
```
2. If I don't put the `output: export` key in my config, then when I `npm build`, everything is "compiled" not in a out/ folder but in the `next/` folder. And this folder structure doesn't tell me what files I have to serve with Nginx, as the only candidate is `next/server/pages` but it doesn't seems to have everything packed up. I think the .next/ folder only works as cache assets, but not still answer to the request; It's not REAL SSG where an HTML file is served and period.

So I don't really understand how I can serve my static files with my webserver and have my node listening behind for orders like page regeneration. Which folder do I have to serve with static HTMLs? If I can't rely on a webserver, and plug directly node on my reverse proxy, how can I be sure that I'm serving static HTML and not doing classic SSR? If ISR is SSR with cache, how can I ensure what will happen when requesting https://whatever.com/article/foo?

Then after a lot of research around how to deploy a NextJS project, I feel like there should be a bit of revamp around the server modes. I feel like the actual SSR/ISR/SSG modes demarcations are suboptimal and can be easily improved.

- SSR works without static HTML files, as HTML is generated on each request.
- ISR needs a build step, and those built files are served in a .next/ folder.
- SSG needs only a build step, and those built files are served in a out/ folder.

So if we look at them from the cache perspective:

- SSG is an ad-vitam cache, static HTML files that won't be modified (until they are),
- ISR is a request with cached front, that's HTML files that can change numerous time,
- SSR has no cache (at least, not front one!), every response is uniquely crafted,

Maybe ISR and SSG should serve their static files in the same folder, then:
- If the folder is empty, or the client requests a non existing file, the node server take the request,
- If the file exists, serve it, with node or with a typical webserver (apache, nginx, caddy, IIS, ...)
- If the file doesn't exist and there is no running node behind (SSG), then request will fail and that's ok.

Then we can get rid of the output: export config parameter?

I would say that we should get rid of out/ and .next/ to always export in public/. For convenience in folder management, maybe move public/assets/ to assets/, at the root. And let Next manage public/, puting the generated HTML files or an entrypoint in it:
- If you use a Nginx and a node server, you can do SSR and ISR, not SSG (well you can but the node server will never be used here),
- If you only use a Nginx server, you can do SSG, no SSR or ISR,
- If you only use a node server, you can do SSR or ISR, no SSG (assuming node will serve cached HTML files for ISR).

Only one structure, everyone choice for the configuration and infrastructure, everything is simplier.

**Comment by goustaff** [comment]
Hello!
I absolutely don't know anything about _PHP_, nor _Traefik_.

I'll try to answer to some points about _Next.js_.
I'm a beginner on this framework, so I hope I'm not wrong.

---

> It's not REAL SSG where an HTML file is served and period

Hmmm...
_Next.js_ needs to process some hydration.
So it will need to serve also chunked _JS_ files.
However, a static web page can include both _HTML_, _CSS_, and _JS_. So I'm not sure to fully understand this point.

Theoretically, if I go back to the documentation:

> If a page uses Static Generation, the page HTML is generated at build time. That means in production, the page HTML is generated when you run next build. This HTML will then be reused on each request. It can be cached by a CDN.

- https://nextjs.org/docs/pages/building-your-application/rendering/static-site-generation

---

> As I need to dynamically manage my number of article pages, I would try to deploy with ISR mode

This involves fetching these posts in order to retrieve them all (for the category and sub-category implied by the post), then counting them in order to generate the pagination. I'm struggling to see what the problem is here, it seems like a good _SSG_ use case to me.

_ISR_ is just _SSG_ with an additional revalidation parameter (_TTL_).

> `revalidate: 10, // In seconds`

- https://nextjs.org/docs/pages/building-your-application/data-fetching/incremental-static-regeneration

So far, it seems like a pretty good approach and I don't see any stress in it.

---

> 1 If I put `output: export` in my _next.config.js_ file, then I can't use npm start anymore:

The error message seems to be explicit enough:

> Use "npx serve@latest out" instead.

However, using `output: export` introduces a much more serious problem in this case, and should be avoided.
It breaks the _ISR_, and many other features:

> Unsupported Features
> - ...
> - Incremental Static Regeneration

- https://nextjs.org/docs/pages/building-your-application/deploying/static-exports

---

> 2. If I don't put the `output: export` key in my config, then when I npm build, everything is "compiled" not in a `out/` folder but in the `.next/` folder. And this folder structure doesn't tell me what files I have to serve with Nginx, as the only candidate is `next/server/pages` but it doesn't seems to have everything packed up.

No. Everything is not packed up.
The JS is indeed chunked, particularly for _hydration_ concerns.

You can explore those chunks in-depth with `@next/bundle-analyzer`, and also optimize them with `dynamic`.
- https://www.npmjs.com/package/@next/bundle-analyzer
- https://nextjs.org/docs/pages/building-your-application/optimizing/lazy-loading

These chunks mechanisms can also mean than "More" than "everything" (you _want_ for _one_ page) can be packed up in the loaded JS, depending on how you isolate your imports, and how you use lazy loading.

---

> How can I be sure that I'm serving static HTML and not doing classic SSR

I think you could just send a _404_ error if the page doesn't exist.

- https://nextjs.org/docs/app/api-reference/functions/not-found

Another way to standardize this would probs be to use the _manifests_ files, I guess?
Some of them are automatically generated by _Next.js_.

- https://github.com/vercel/next.js/discussions/14779
- https://nextjs.org/docs/app/api-reference/file-conventions/metadata/manifest

Maybe you could also dig in the code of `next-sitemap` to understand how to retrieve the built _SSG_ paths:
- https://github.com/iamvishnusankar/next-sitemap
(Attention: there is currently some very odd issues with this lib.)

Also: in an _SSG_ context, _SSR_ will only be triggered for unknown _Static Params_. All the _Static Params_ you define in the `generateStaticParams` will trigger _SSG_, and not _SSR_.

---

> Which folder do I have to serve with static HTMLs?

Maybe folders/files like:
`.next/server/app/en/dashboard/foo.html` ?

But in the end, these files are the same ones that will be served by _Next.js_ to the corresponding _URLs_, so I don't really understand?

---

> If ISR is SSR with cache

_ISR_ is not _SSR_ with cache.
_SSG_ is _SSR_ with cache, and _ISR_ is just _SSG_ with a regenerate _TTL_ (seconds) parameter.

Also, notice this _TTL_ pattern can also be used with `fetch` in _Next.js_, with the same logic:
```tsx
fetch('https://...', { next: { revalidate: 3600 } })
``` 
- https://nextjs.org/docs/app/building-your-application/data-fetching/fetching-caching-and-revalidating#revalidating-data

Likewise:
- The "R" of _SSR_ (Server-side Rendering) means _Rendering_,
- Whereas the "R" of _ISR_ (Incremental Static Regeneration) means _Regeneration_.

Then, _SSR_ and _ISR_ are absolutely not for the same use cases (I also find it confusing).

So:
- _SSR_ means "Page regenerated on each request",
- Whereas _ISR_ means "Page generated at build time, then regenerated according to a _TTL_" (_SSG_ with _TTL_).

---

> how can I ensure what will happen when requesting https://whatever.com/article/foo?

I don't really understand this question: maybe you would like to trigger a _404_, or a redirection if the dynamic path isn't valid?

---

> I would say that we should get rid of out/ and .next/ to always export in public/. For convenience in folder management, maybe move public/assets/ to assets/, at the root. And let Next manage public/, puting the generated HTML files or an entrypoint in it: [...]

I think it is a too big breaking change, which would break all the ecosystem.
I won't go into more detail on these considerations.

Finally, maybe what you're looking for is a lib like _ContentLayer_?
- https://contentlayer.dev/

I hope I've at least cleared up a few ambiguities.

---

EDIT (08/11, 14h50): There were some important points missing from my answer.
I look forward to hearing from you again.

EDIT [2] (09/11, 17h26): _TTL_\*, not _TTY_.

---

P.S.: To be honest, _I think_ I'd like to take the discussion back to where it started, even before your first post.
I mean, to really understand what the initial need is, and then reintroduce the stack considerations. All of this looks like a _XY problem_.
